### PR TITLE
Allow user to custom response content-type via options

### PR DESCRIPTION
**Pull request by janRucka**
Hi,

I would like to have an option to configure response content-type. 

My use case is REST endpoint used for export. Export format is determined by Accept header (either csv or xml). In case of error I want to always return JSON to be consistent (and I don't want to force clients to parse error details from csv/xml).

This change would make my life easier and I think others could benefit from that as well. Hope you will include it.

### CODE REVIEW

**Comment by khellang** [comment]
I'm not sure how this is supposed to work. [RFC7807](https://tools.ietf.org/html/rfc7807) only specifies `application/problem+json` and `application/problem+xml` as valid media types. I don't think we should drop the spec and try rolling our own formats for this.

**Comment by janRucka** [comment]
There are two reasons for the change. First I already described. In case of error I want to always return JSON. But if my export endpoint exports into XML how can I do that? Client will send request with XML accept header which results into error in XML not in JSON (in my case client is currently unable to parse XML errors). Maybe there is some way around I don't know about. If so please let me know.

Second the specification. I didn't read it thoroughly but I don't really agree with you. You are right that it specifies only `application/problem+json` and `application/problem+xml` but it also says that you can specify your own format in 4th paragraph "Defining New Problem Types". Here is quote from that:

> When an HTTP API needs to define a response that indicates an error
>    condition, it might be appropriate to do so by defining a **new problem**
>    **type**.

> Finally, an **application might have a more appropriate way to carry an**
>    **error in a format that it already defines**.  Problem details are
>    intended to avoid the necessity of establishing new "fault" or
>    "error" document formats, not to replace existing domain-specific
>    formats.

**Comment by khellang** [comment]
> I didn't read it thoroughly but I don't really agree with you.

Maybe read it properly first? 😅 

> it also says that you can specify your own format in 4th paragraph "Defining New Problem Types"

Yes, but you conveniently left out the entire remaining part of that section, including [Section 4.1. Example](https://tools.ietf.org/html/rfc7807#section-4.1):

>  For example, if you are publishing an HTTP API to your online
>  shopping cart, you might need to indicate that the user is out of
>  credit (our example from above), and therefore cannot make the
>  purchase.
>
>  If you already have an application-specific format that can
>  accommodate this information, it's probably best to do that.
>  However, if you don't, you might consider using one of the problem
>  details formats -- JSON if your API is JSON-based, or XML if it uses
>  that format.

A "Problem Type" isn't a different media type format. It's referring to the `type` of the problem returned, in either XML or JSON.

If you really want to change the `ContentTypes` of the problem details type, you should use the existing `OnBeforeWriteDetails` option instead.

**Comment by janRucka** [comment]
It seems that you are right type in this context really mean just member as described in 3.1. Honestly I am a bit confused with the terminology. But I still think you can have your own format (that is how they call it):

> This specification defines simple JSON [RFC7159] and XML
>    [W3C.REC-xml-20081126] document **formats** to suit this purpose.

>    Note that problem details are (naturally) not the only way to convey
>    the details of a problem in HTTP; if the response is still a
>    representation of a resource, for example, it's often preferable to
>    accommodate describing the relevant details in that application's
>    **format**.

>  Finally, an application might have a more appropriate way to carry an
>    error in a **format** that it already defines.  Problem details are
>    intended to avoid the necessity of establishing new "fault" or
>    "error" document **formats**, not to replace existing domain-specific
>    **formats**.

---------------------

> If you really want to change the `ContentTypes` of the problem details type, you should use the existing `OnBeforeWriteDetails` option instead.

You can add `ContentTypes` but you cannot change them or remove the default ones (well if you could then my change would indeed be pointless). I would need something like `OnBeforeExecute`.

-------------------

I wanted to make my changes more general but I really don't want to argue about interpretation of some standard. What I need is to have REST endpoint for export. Where export format is determined by Accept header (either csv or xml). In case of error I want to always return JSON to be consistent (and I don't want to force clients to parse error details from csv/xml). I don't think it is possible to do right now (I wrapped ProblemDetails middleware but it is not very nice solution). How this should be done?

**Comment by janRucka** [comment]
@khellang: Did you read my previous comment? What do you think?

**Comment by janRucka** [comment]
@khellang I don't want to be inpatient but it is more than two weeks now.

**Comment by khellang** [comment]
Hmm, I'm not a fan of forcing a media format that the client doesn't accept.

Ideally, the server should return back a `406 Not Acceptable` status, but since this is for exceptional cases, I think it's OK to return a format that the client doesn't accept, like returning JSON if the client requested `text/html`.

However, if the client explicitly says it accepts XML, we should probably follow that request, no?

**Comment by janRucka** [comment]
In my case client doesn’t support XML it is used just to specify format for export (which is file for user). And as far as I know this is valid way to do it by specifying Accept header (opposite way would be to have separate endpoints). That is my use case, I am not able to consume errors in XML and I want to be consistent and always return error in JSON. And I think it is valid use case.

On more general note I think that client is not necessarily able to receive errors in specified format. In some cases (e.g. "application/octet-stream") it is hardly imaginable. I don’t want to force client to be able parse error in several formats (JSON for regular endpoints and XML for export endpoints). If regular endpoints would return XML, then it makes sense to return XML error but this is not the case.

The change doesn’t force anybody to do anything. It doesn’t change anything by default. It just allows someone (like me) to easily configure the behavior as they need. The use case is probably not that common but still I think it is nice to have. Also, it could help for rare cases when someone needs different error format. 

I can fully understand that you don’t want to merge any nonsensical changes from random people. But I think this change has it’s valid (although rare) use cases. If you don’t like the implementation, I can change it if you give me some direction.

File: src/ProblemDetails/ProblemDetailsMiddleware.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
Comment (khellang) [inline src/ProblemDetails/ProblemDetailsMiddleware.cs]: 🔥

File: src/ProblemDetails/ProblemDetailsOptions.cs [hunk 2]
```diff
@@ -46,6 +49,8 @@ public ProblemDetailsOptions()
 
         private List<ExceptionMapper> Mappers { get; }
 
+        public MediaTypeCollection ResponseContentType { get; }
```
Comment (khellang) [inline src/ProblemDetails/ProblemDetailsOptions.cs]: This should probably be `ResponseContentTypes` to match the `ContentTypes` property it's being used for.

File: src/ProblemDetails/ProblemDetailsOptions.cs [hunk 3]
```diff
@@ -26,6 +27,8 @@ public ProblemDetailsOptions()
 
                 HeaderNames.WWWAuthenticate,
             };
+
+            ResponseContentType = new MediaTypeCollection();
```
Comment (khellang) [inline src/ProblemDetails/ProblemDetailsOptions.cs]: Can you move this above `AllowedHeaderNames`?

**Comment by janRucka** [comment]
I fixed all your notes.

Ad unit tests: Unfortunately I have troubles to make TestServer to return anything else than JSON. Not sure what is wrong and right now I don't have time to investigate it deeper.

**Comment by khellang** [comment]
> Unfortunately I have troubles to make TestServer to return anything else than JSON.

You probably have to do some work to be able to configure this;

https://github.com/khellang/Middleware/blob/7e4943b6153e2ffb4eabe3b53b4bede3345cb54f/test/ProblemDetails.Tests/ProblemDetailsMiddlewareTests.cs#L314-L315

File: src/ProblemDetails/ProblemDetailsOptions.cs (hunk 2 above)
Comment (khellang) [reply src/ProblemDetails/ProblemDetailsOptions.cs]: I don't think this has been resolved. It's now `ContentType` and not `ContentTypes` 😄

**Comment by janRucka** [comment]
I tried to add other content types there (and even delete JSON). Also added xml formaters (tried to get at least XML). Also explicitly set `options.ReturnHttpNotAcceptable = false;` (which should be default value anyway). But in case of exception from TestServer I got 406...

File: src/ProblemDetails/ProblemDetailsOptions.cs (hunk 2 above)
Comment (janRucka) [reply src/ProblemDetails/ProblemDetailsOptions.cs]: You are right, thanks. It is fixed now.

**Comment by janRucka** [comment]
Sorry for the delay. Tests finally added. My problem turned out to be caused by this issue:
https://github.com/khellang/Middleware/issues/66
unfortunately I spend quite some time looking what is wrong with my setting of WebHost which turned to be fine.

**Comment by janRucka** [comment]
I forced push my conflict resolution (I did rebase since merge seemed messy) and I think changes which you requested were done (if not please let me know).

So everything should be resolved now. Is there a way to re-run the check?

**Comment by khellang** [comment]
> So everything should be resolved now. Is there a way to re-run the check?

The checks should run automatically whenever there's a push to the branch (doesn't matter if it's a rebased force push or not), but there seems to be some merge conflicts still...

**Comment by khellang** [comment]
I think I managed to resolve the conflict in the GitHub UI on my phone 😂 I'll have a proper review tomorrow. Hopefully we can get this in 😉

**Comment by janRucka** [comment]
@khellang Any update?

**Comment by khellang** [comment]
> @khellang Any update?

Not really. There are some conflicts that needs to be resolved. I've also bumped the major version because of some breaking changes (target `netcoreapp3.1` only). If we want to include this in the 4.x version, we need to backport it once the conflicts have been resolved and the PR has been merged.

**Comment by janRucka** [comment]
Should be resolved now. I somehow missed you comment from March 1st.

**Comment by khellang** [comment]
Beautiful! Thanks 😘 Are you interested in me backporting it to a 4.x version or is it enough release a new 5.x beta?

**Comment by khellang** [comment]
OK, I backported it to the 4.x branch and pushed both v5.0.0-beta.4 and v4.4.0 to NuGet.org 😄

**Comment by janRucka** [comment]
Thank you very much!

**Pull request merged**